Add doc Convert an existing non-TLS etcd cluster to TLS

adeptvin1 opened the pull request:
Description:
In some case engineers need to convert existing etcd cluster to TLS communication. Unfortunately, there is almost no documentation in the public domain. I found a very old instruction, but it seems to me that it would be right to put an updated version in the main documentation. I checked, the method works, the conversion is adequate.

Old doc -> https://github.com/coreos/docs/blob/master/etcd/etcd-live-http-to-https-migration.md

If I need to change something, let me know, thanks!

spzala reviewed (commented):
@adeptvin1 thank you! Please sign your commit to pass DCO.

adeptvin1 commented:
@spzala good day!
I did it thanks!

spzala commented:
Thanks @adeptvin1 I agree that the use case you have mentioned is valid, and the link you have provided is old and it makes sense to add updates as you have provided. Your changes lgtm but I haven't tested your changes. I will wait for more reviews.

spzala commented:
/ok-to-test

ahrtr commented on content/en/docs/v3.6/op-guide/security.md:
````diff
@@ -239,6 +239,98 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+
+Update each etcd member’s peer URL to use `https://`:
+
+```sh
+etcdctl member update 10f4390f4a904a13 --peer-urls=https://ip_address_etcd_node_1:2380
+etcdctl member update 18e03f3590bf045c --peer-urls=https://ip_address_etcd_node_2:2380
+etcdctl member update ca7a7d6fe0984fd9 --peer-urls=https://ip_address_etcd_node_3:2380
+```
````
Isn't this a chick-egg issue?  Once you update the peerURL for each member, then its peers will lose connection with it.

I think we should rolling update the members: update the members one by one. Specifically, repeat the following steps for each member, and verify the health before moving on the next member,
- prepare/copy the certificate/key to the member;
- update peerURL for the member using command `etcdctl member update ...`;
- update `--listen-peer-urls`, `--listen-client-urls`, `--initial-advertise-peer-urls`, `--advertise-client-urls` and `--initial-cluster`;
- restart the member
- verify health & status of the cluster

adeptvin1 replied:
About it)
If you will update --peer-urls= and will restart each etcd node cluster your cluster will go down, because etcd will see only new endpoints (or you should add both, http and https interfaces on different ports)
In my case, I just update peer-urls (etcd cluster looks fine and after that restart etcd systemd service)

ahrtr replied:
> If you will update --peer-urls= and will restart each etcd node cluster your cluster will go down

- Not sure we are on the same page. I was saying to rolling update the members: update one member at a time, and move on to the next member after verifying the health of the cluster;
- We might need to disable (remove) the option `--peer-client-cert-auth=true` during the process, and add it back once we have finished updating all members.




> In my case, I just update peer-urls (etcd cluster looks fine and after that restart etcd systemd service)

It's weird. As I mentioned above, once updating a member's peerURLs, its peers will lose connect with it. @ArkaSaha30 @henrybear327 are you able to reproduce/verify it? thx

henrybear327 replied:
I will try to follow the documentation and reproduce it later today!

henrybear327 replied:
I haven't gotten all the way through yet, but if I only go up till `member update`, my cluster seems to continue to work.

Here is what I have done:

### Create a cluster 
```sh
# cluster 1c45a069f3a1d796

# e1 b71f75320dc06a6c
./bin/etcd \
            --name e1 \
            --initial-advertise-peer-urls http://127.0.0.1:2380 \
            --listen-peer-urls http://127.0.0.1:2380 \
            --advertise-client-urls http://127.0.0.1:2379 \
            --listen-client-urls http://127.0.0.1:2379 \
            --initial-cluster "e1=http://127.0.0.1:2380" \
            --initial-cluster-state new 

# e2 5ba6f5b74447021e
./bin/etcdctl member add e2 \
            --peer-urls=http://127.0.0.1:2382 \
            --learner
./bin/etcd \
            --name e2 \
            --initial-advertise-peer-urls http://127.0.0.1:2382 \
            --listen-peer-urls http://127.0.0.1:2382 \
            --advertise-client-urls http://127.0.0.1:2378 \
            --listen-client-urls http://127.0.0.1:2378 \
            --initial-cluster "e1=http://127.0.0.1:2380,e2=http://127.0.0.1:2382" \
            --initial-cluster-state existing
./bin/etcdctl member promote 5ba6f5b74447021e

# e3 9cd1421e057683d3
./bin/etcdctl member add e3 \
            --peer-urls=http://127.0.0.1:2384 \
            --learner
./bin/etcd \
            --name e3 \
            --initial-advertise-peer-urls http://127.0.0.1:2384 \
            --listen-peer-urls http://127.0.0.1:2384 \
            --advertise-client-urls http://127.0.0.1:2377 \
            --listen-client-urls http://127.0.0.1:2377 \
            --initial-cluster "e1=http://127.0.0.1:2380,e2=http://127.0.0.1:2382,e3=http://127.0.0.1:2384" \
            --initial-cluster-state existing
./bin/etcdctl member promote 9cd1421e057683d3
```

### Cluster check
```sh
./bin/etcdctl --endpoints=http://127.0.0.1:2377,http://127.0.0.1:2378,http://127.0.0.1:2379 member list
```

```
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl --endpoints=http://127.0.0.1:2377,http://127.0.0.1:2378,http://127.0.0.1:2379 member list
5ba6f5b74447021e, started, e2, http://127.0.0.1:2382, http://127.0.0.1:2378, false
9cd1421e057683d3, started, e3, http://127.0.0.1:2384, http://127.0.0.1:2377, false
b71f75320dc06a6c, started, e1, http://127.0.0.1:2380, http://127.0.0.1:2379, false
```

### Start to perform member update

```sh
./bin/etcdctl member update b71f75320dc06a6c --peer-urls=https://127.0.0.1:2380
./bin/etcdctl --endpoints=http://127.0.0.1:2377,http://127.0.0.1:2378,http://127.0.0.1:2379 member list
```

Response
```
Member b71f75320dc06a6c updated in cluster 1c45a069f3a1d796
```

#### e1 log

```
{"level":"info","ts":"2025-04-09T20:46:11.578822Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1981","msg":"b71f75320dc06a6c switched to configuration voters=(6604236071058735646 11299885636757390291 13195394291058371180)"}
{"level":"info","ts":"2025-04-09T20:46:11.578888Z","caller":"membership/cluster.go:567","msg":"updated member","cluster-id":"1c45a069f3a1d796","local-member-id":"b71f75320dc06a6c","updated-remote-peer-id":"b71f75320dc06a6c","updated-remote-peer-urls":["https://127.0.0.1:2380"],"updated-remote-peer-is-learner":false}
{"level":"info","ts":"2025-04-09T20:46:11.578914Z","caller":"etcdserver/server.go:1751","msg":"applied a configuration change through raft","local-member-id":"b71f75320dc06a6c","raft-conf-change":"ConfChangeUpdateNode","raft-conf-change-node-id":"b71f75320dc06a6c"}
```

#### e2 log

```
{"level":"info","ts":"2025-04-09T20:46:11.579061Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1981","msg":"5ba6f5b74447021e switched to configuration voters=(6604236071058735646 11299885636757390291 13195394291058371180)"}
{"level":"info","ts":"2025-04-09T20:46:11.579103Z","caller":"membership/cluster.go:567","msg":"updated member","cluster-id":"1c45a069f3a1d796","local-member-id":"5ba6f5b74447021e","updated-remote-peer-id":"b71f75320dc06a6c","updated-remote-peer-urls":["https://127.0.0.1:2380"],"updated-remote-peer-is-learner":false}
{"level":"info","ts":"2025-04-09T20:46:11.579123Z","caller":"rafthttp/transport.go:390","msg":"updated remote peer","local-member-id":"5ba6f5b74447021e","updated-remote-peer-id":"b71f75320dc06a6c","updated-remote-peer-urls":["https://127.0.0.1:2380"]}
{"level":"warn","ts":"2025-04-09T20:46:16.579916Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:16.579981Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:21.581550Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:21.581521Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:26.583160Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:26.583226Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:31.583380Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:31.583400Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:36.585008Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:36.585096Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:41.587195Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:41.587247Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:46.590041Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:46.590169Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:51.591340Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:51.591385Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:56.596107Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:56.596141Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
```

#### e3 log

```




{"level":"info","ts":"2025-04-09T20:46:11.579259Z","logger":"raft","caller":"v3@v3.6.0/raft.go:1981","msg":"9cd1421e057683d3 switched to configuration voters=(6604236071058735646 11299885636757390291 13195394291058371180)"}
{"level":"info","ts":"2025-04-09T20:46:11.579298Z","caller":"membership/cluster.go:567","msg":"updated member","cluster-id":"1c45a069f3a1d796","local-member-id":"9cd1421e057683d3","updated-remote-peer-id":"b71f75320dc06a6c","updated-remote-peer-urls":["https://127.0.0.1:2380"],"updated-remote-peer-is-learner":false}
{"level":"info","ts":"2025-04-09T20:46:11.579324Z","caller":"rafthttp/transport.go:390","msg":"updated remote peer","local-member-id":"9cd1421e057683d3","updated-remote-peer-id":"b71f75320dc06a6c","updated-remote-peer-urls":["https://127.0.0.1:2380"]}
{"level":"warn","ts":"2025-04-09T20:46:16.579917Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:16.579998Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:21.581535Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:21.581635Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:26.583421Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:26.583490Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:31.584333Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:31.584378Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:36.584941Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:36.585025Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:41.587299Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:41.587337Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:46.589921Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:46.590053Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:51.590267Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:51.590295Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:56.596107Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:46:56.596143Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:01.597130Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:01.597168Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:06.598816Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:06.598863Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:11.600501Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_RAFT_MESSAGE","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
{"level":"warn","ts":"2025-04-09T20:47:11.600581Z","caller":"rafthttp/probing_status.go:68","msg":"prober detected unhealthy status","round-tripper-name":"ROUND_TRIPPER_SNAPSHOT","remote-peer-id":"b71f75320dc06a6c","rtt":"0s","error":"tls: first record does not look like a TLS handshake"}
```

Performing some put and get operations seem to work though:
```sh
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl put a 1 -w json
{"header":{"cluster_id":2037210783374497686,"member_id":13195394291058371180,"revision":2,"raft_term":2}}
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl put b 2 -w json
{"header":{"cluster_id":2037210783374497686,"member_id":13195394291058371180,"revision":3,"raft_term":2}}
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl get "" --prefix=true
a
1
b
2
```

If I now do to update e2, too (now 2 out of 3 nodes are updated)
```sh
./bin/etcdctl member update 5ba6f5b74447021e --peer-urls=https://127.0.0.1:2382
```

The `tls: first record does not look like a TLS handshake` will now occur in all nodes' log. But the put and get is still working.
```sh
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl put a 1 -w json
{"header":{"cluster_id":2037210783374497686,"member_id":13195394291058371180,"revision":11,"raft_term":2}}
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl put b 2 -w json
{"header":{"cluster_id":2037210783374497686,"member_id":13195394291058371180,"revision":12,"raft_term":2}}
vscode ➜ /workspaces/etcd (main) $ ./bin/etcdctl get "" --prefix=true
a
1
b
2
```

henrybear327 replied:
Not sure if this is enough information for you @ahrtr.

ahrtr replied:
Thanks @henrybear327 for the update. 

After you updated the second member's peerURL, if you restarts all members, then you might not be able to read/write data anymore (But if you connect the leader,  it should still work). Also when you continue to update third member's peerURL, it should fail. But it still works if we don't restart the members. It'd be great if you or anyone could figure out the root cause, although we don't have to necessarily to fix it.
- I do not get time to dig into the root cause, the reason could be that each member is still listening on the old peerURL until it gets restarted.


I still think we should rolling updating each member as mentioned above. can anyone verify the steps? thx

adeptvin1 replied:
I may be out of place, but it also seems to me that before reboot it listens to old endpoints, so in our implementation of ansible playbook we do exactly as I described above. First we update members and then reboot the systemd service (we have a deployment on systemd )

We checked on several combat etcd with patroni and everything works fine.

ahrtr replied:
> But it still works if we don't restart the members. It'd be great if you or anyone could figure out the root cause, although we don't have to necessarily to fix it.
> 
> * I do not get time to dig into the root cause, the reason could be that each member is still listening on the old peerURL until it gets restarted.

To be clearer, the steps (update all members' peerURLs at the same time) are supposed to be not working, but somehow it works. We need to figure out the root cause, and decide whether we should fix it later. @henrybear327 I am happy to discuss this if you have time.

adeptvin1 replied:
@henrybear327 good day.
If you're going to look at the code anyway, could you please look at this feature? It seems to me that this is a mistake by previous colleagues that can be easily fixed. https://github.com/etcd-io/etcd/discussions/19734
This is a question just within the framework of conversion to TLS etcd cluster.

Thanks

ahrtr commented on content/en/docs/v3.6/op-guide/security.md:
```diff
@@ -239,6 +239,98 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+
+### 3. Enable TLS for client-to-server communication
+
+Edit the systemd unit file (`/etc/systemd/system/etcd.service`) or drop-in override config on each node to add the following flags:
```
minor comment: suggest not to mention the `/etc/systemd/system/etcd.service`, we just need to clarify which etcd options need to be updated.

adeptvin1 replied:
I'll update it

adeptvin1 commented on content/en/docs/v3.6/op-guide/security.md:
````diff
@@ -239,6 +239,98 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+--peer-trusted-ca-file=/etc/ssl/etcd/ca.pem
+```
+
+Also, update all `--listen-peer-urls`, `--listen-client-urls` and `--initial-advertise-peer-urls` to use `https://`.
````
I'll update it

adeptvin1 replied:
@ahrtr i did it, can you check?

adeptvin1 commented:
Good day!
Do you have any updates? Maybe I need to do something?
Thanks!

ahrtr commented:
As we discussed in https://github.com/etcd-io/website/pull/977#discussion_r2032660277, normally it works to update all members' peerURLs at a time because each member still listens on the old peerURLs until it gets restarted or the old connections somehow break (not 100% sure about this).

However this approach seems have flaws. What if the member crashes (and restarted) right after you update its peerURLs? Or what if other peer(s) crashes (and restarted) right after you update a member's peerURLs? Please try it out. My draft thought is that a rolling update might be the most prudent (but might be a little more boring) approach.

All guides provided on etcd.io should be official & carefully verified. We need someone to take ownership on this to get all the confusion sorted out.

```sh
etcdctl member update 10f4390f4a904a13 --peer-urls=https://ip_address_etcd_node_1:2380
etcdctl member update 18e03f3590bf045c --peer-urls=https://ip_address_etcd_node_2:2380
etcdctl member update ca7a7d6fe0984fd9 --peer-urls=https://ip_address_etcd_node_3:2380
```

jberkus commented:
I'm going to add this PR to an upcoming community meeting.  @adeptvin1 if you can make the next meeting, you could answer questions about it.

adeptvin1 commented:
Good day!
No problem, I'll do it today.

adeptvin1 commented:
@ahrtr Good day!
I repeated your plan -> https://github.com/etcd-io/website/pull/977#discussion_r2032660277.

- prepare/copy the certificate/key to the member (Done for all);
- update peerURL for the member using command etcdctl member update ... (Done for all);
- update --listen-peer-urls, --listen-client-urls, --initial-advertise-peer-urls, --advertise-client-urls and --initial-cluster (Done for all, Cluster Still working);
restart the member (I did it on 1 member, systemd restart was stuck)
verify health & status of the cluster (Cluster is unhealthy)

```sh
etcdctl endpoint health --user=blablabla --password=blablabla--cluster
{"level":"warn","ts":"2025-06-06T14:09:51.99664+0300","logger":"client","caller":"v3@v3.5.10/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0003468c0/192.168.4.2:2379","attempt":0,"error":"rpc error: code = DeadlineExceeded desc = context deadline exceeded"}
http://192.168.6.2:2379 is healthy: successfully committed proposal: took = 2.131833ms
http://192.168.3.2:2379 is healthy: successfully committed proposal: took = 5.221961ms
http://192.168.4.2:2379 is unhealthy: failed to commit proposal: context deadline exceeded
```

I repeat it for one member

- prepare/copy the certificate/key to the member (Done for all);
- update peerURL for the member using command etcdctl member update ... (Done for all);
- update --listen-peer-urls, --listen-client-urls, --initial-advertise-peer-urls, --advertise-client-urls and --initial-cluster (Done for 1 member, Cluster Still working);
restart the member (I did it on 1 member, systemd restart was ok)

```sh
85+0300","caller":"embed/config_logging.go:169","msg":"rejected connection","remote-addr":"192.168.3.2:39300","server-name":"","error":"remote error: tls: bad certificate"}
```

- verify health & status of the cluster (Cluster is unhealthy)


I think that when we have restart "member crashes (and restarted) right after you update its peerURLs" we should just restart another node for continue work etcd cluster.

In this case i think we should change doc. Like this -> https://github.com/coreos/docs/blob/master/etcd/etcd-live-http-to-https-migration.md
Move "Enable TLS for client-to-server communication"  on point 2 and after that update member list. Whenever etcd cluster was broken if we have a "member crashes (and restarted)" we can just update member list and restart another members.

adeptvin1 commented:
Hello.
I think I wrote that you can continue working with etcd in case of a reboot/problems with 1 of the nodes. The question is that there should be at least 2 of them. If we convert a cluster of 3 nodes one by one, then at the first stage we will have 1 node with TLS, 2 without TLS and the cluster will respond without TLS. At the 2nd stage we convert the 2nd node, we will get a cluster with TLS (2 out of 3 with TLS) and it will also work.

Or what is the question?
As a last resort, you can add a block about taking a backup from the etcd cluster for maximum security.

ahrtr commented:
> If we convert a cluster of 3 nodes one by one, then at the first stage we will have 1 node with TLS, 2 without TLS and the cluster will respond without TLS. At the 2nd stage we convert the 2nd node, we will get a cluster with TLS (2 out of 3 with TLS) and it will also work.

It looks good. Pls update the doc.

adeptvin1 commented:
@ahrtr 
Greetings.
I added a step so that colleagues would definitely add TLS certificates (just in case), added an addition regarding node reboot as we discussed earlier. Added node reboot to the member update commands as part of our discussion. Please take a look, will this be better?

ahrtr commented on content/en/docs/v3.6/op-guide/security.md:
````diff
@@ -239,6 +239,105 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+
+Please add TLS certificates to each node using the instructions above.
+
+### 3. Enable TLS for client-to-server communication
+
+Update each node's parameters to match your etcd deployment method using the following flags:
+
+```ini
+--cert-file=/etc/ssl/etcd/server.pem
+--key-file=/etc/ssl/etcd/server-key.pem
+--client-cert-auth=true
+--trusted-ca-file=/etc/ssl/etcd/ca.pem
+--peer-cert-file=/etc/ssl/etcd/peer.pem
+--peer-key-file=/etc/ssl/etcd/peer-key.pem
+--peer-client-cert-auth=true
+--peer-trusted-ca-file=/etc/ssl/etcd/ca.pem
+```
+
+Also, update all `--listen-peer-urls`, `--listen-client-urls`, `--advertise-client-urls`, `--initial-cluster`  and `--initial-advertise-peer-urls` to use `https://`.
````
After step 3 is done, it would be better to restart all etcd instances, and verify the client-server communication is working as expected before users go ahead to continue to update peer-to-peer communication.

adeptvin1 replied:
But if we reboot the nodes before doing member update etcd the node will hang. I checked this earlier.

adeptvin1 replied:
Do I understand correctly that you propose to split the step into two?
First, update the connection between etcd nodes (what you specified is responsible for mTLS between nodes) and only after that update the connection to TLS for the peer?

Below I have a comment about updating urls from http to https, I can move it to the main block where the TLS parameters are.

jberkus commented:
Where are we with this doc?  @ahrtr @adeptvin1 ?

adeptvin1 commented on content/en/docs/v3.6/op-guide/security.md:
````diff
@@ -239,6 +239,105 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+Update each etcd member’s peer URL to use `https://` step by step with restart:
+
+```sh
+$ On node 10f4390f4a904a13
+etcdctl member update 10f4390f4a904a13 --peer-urls=https://ip_address_etcd_node_1:2380
+systemctl daemon-reload
+systemctl restart etcd
+$ On node 18e03f3590bf045c
+etcdctl member update 18e03f3590bf045c --peer-urls=https://ip_address_etcd_node_2:2380
+systemctl daemon-reload
+systemctl restart etcd
+$ On node ca7a7d6fe0984fd9
+etcdctl member update ca7a7d6fe0984fd9 --peer-urls=https://ip_address_etcd_node_3:2380
+systemctl daemon-reload
+systemctl restart etcd
````
@ahrtr good day!
what I need to do?

ahrtr commented:
pls squash the commits, and ensure it's signed off

jberkus commented:
OK, lemme give it a re-read for grammar etc.

jberkus commented on content/en/docs/v3.6/op-guide/security.md:
```diff
@@ -239,6 +239,107 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+Previously, we described that etcd supports two types of communication: **client-to-server** and **server-to-server (peer communication)**. In this section, we will prepare configuration steps to migrate both communication types to TLS in your existing etcd cluster.
+
+> **Important**: Before starting work on converting the cluster to TLS, please make a backup of the etcd database.
+> **Important**: Before restarting any etcd members, you must first add the TLS certificates and configuration to each etcd node.
```
add a line break between these so that they're two separate warnings

```suggestion

> **Important**: Before restarting any etcd members, you must first add the TLS certificates and configuration to each etcd node.
```

adeptvin1 replied:
@jberkus ^

ivanvc replied:
We could disable MD028. Ref: https://github.com/DavidAnson/markdownlint/blob/main/doc/md028.md

jberkus commented on content/en/docs/v3.6/op-guide/security.md:
```diff
@@ -239,6 +239,107 @@ To disable certificate chain checking, invoke curl with the `-k` flag:
+
+### 2. Add TLS certs on each node
+
+Please add TLS certificates to each node using the instructions above.
```
above where?  Please reference a section so that the reader knows how far to scroll.

adeptvin1 replied:
@jberkus ^

jberkus reviewed (changes requested):
Two minor writing issues, please fix.

adeptvin1 commented:
@ahrtr @jberkus 
Good day. Who should close the requests? When can we merge the changes into the documentation?)

jberkus commented:
Let's merge as-is.  I suspect that we really need to use something other than blockquotes for callouts, but that would become a completely separate PR.
